### DISCUSSION TITLE
Revamp with Component-based framework

### DISCUSSION THREAD

**Discussion by SavvasStephanides** [open-sdg/open-sdg#1443] [Ideas]
Our platform will likely benefit from a complete revamp with a component-based framework such as [React](https://reactjs.org/).

**Comment by brockfanning** [comment]
This is a great idea, and would definitely make future development easier. I don't have much experience with this sort of thing, but I have heard great things about Vue.js. The Gatsby/React combo is popular as well. And of course it would be interesting to see how much can be done using Web Components.

I know it's super early to be talking details, but one high-level requirement that would be good to keep in mind: a strong feature of Open SDG is the ability to use core Open SDG but still override specific components. Right now we're doing this by virtue of being a Jekyll theme. Whatever direction we go in here, it would be important to keep this ability to override when needed.

**Comment by SavvasStephanides** [comment]
Thanks Brock.

I think React will be more than enough for our needs. It's fully component based, which gives the site the modularity it needs for anyone to override whenever needed. Using additional components such as Gatsby might be a bit of an overkill since it's mostly for building blogs (similar to Jekyll).

The only downside I can think of is that React has a bit of a steep learning curve but once you get the hang of it, it'll be a life saver.

**Comment by brockfanning** [comment]
Yes, ease of overriding components is important. If overriding a React component is harder than overriding a Jekyll "include", then that's definitely a concern.

Question - do you see this revamp as layering React (or Vue or web components) on top of Jekyll? Or do you see this as a move away from Jekyll?

**Comment by SavvasStephanides** [comment]
I don't thing component overriding will be a concern. I think it'll be much easier to override and manage components in React than it is with Jekyll. 

> Question - do you see this revamp as layering React (or Vue or web components) on top of Jekyll? Or do you see this as a move away from Jekyll?

I am seeing this as standalone React and a move away from Jekyll for multiple reasons.

**Comment by brockfanning** [comment]
A couple concerns there:

In regards to overrides: Do you have any idea what the mechanism for overriding components might be? Just like with the current Jekyll setup, we want to make sure that it's possible for countries use a "core" Open SDG, and still override components. Put another way, we don't want each country to be maintaining ALL components in their repositories - instead we want them to only maintain those components they're overriding. This is important so that Open SDG can continue to be used as a package/library, instead of as a forkable starter project.

In regards to standalone React: I think it's important to keep in mind our current requirements for technical knowledge, and try to keep things roughly at the same level. For example, right now if a country wants to add a page to the site, it might be something like:
```
---
title: My title
---
# My section header
My paragraph, etc...
```
What would that be in standalone React? I would be nervous if it were something like:
```
const ReactMarkdown = require('react-markdown/with-html')

const markdown = `
# My section header
My paragraph, etc...
`

<ReactMarkdown
  source={markdown}
  escapeHtml={false}
/>
```
I know that something like this would probably be confusing for some of the teams using Open SDG right now. I feel like we would need some layer in between the users and React - like Jekyll or Gatsby (more likely Gatsby, if we use React, right?).

**Comment by SavvasStephanides** [comment]
> Do you have any idea what the mechanism for overriding components might be? 

I haven't figured out the details yet but the way I have it in my head currently is something like this. I'm using the recent example  of updating our contrast buttons to be fully labelled and accessible:

So suppose we have a heading as a component (called `Heading`). It will be part of the page like so:

```js
class Page extends React.Component{
    render(){
        return(
            <div className="Page">
                <Heading/>
                {/* Other stuff */}
            </div>
        )
    }
}
```

The `Heading` component will contain the component for our contrast buttons with the rather uninspiring name of `ContrastButtons`:

```js
class Heading extends React.Component{
    render(){
        return(
            <div className="heading">
                {/* Logo component*/}
                {/* Menu component */}
                <ContrastButtons/>
            </div>
        )
    }
}
```

Now suppose the `ContrastButtons` component contains the buttons for contrast, which are labelled `A`.

```js
class ContrastButtons extends React.Component{
    render(){
        return(
            <div className="contrastButtons">
                <a className="highContrastButton" href="#setToHightContrast">A</a>
                <a className="defaultContrastButton" href="#setToLowContrast">A</a>
            </div>
        )
    }
}
```

Now for accessibility reasons, we decide to have fully labelled buttons, but we want to make them optional. We can add an optional `prop` for devs to opt in (or out) of the new labels:

```js
class Heading extends React.Component{
    render(){
        return(
            <div className="heading">
                {/* Logo component*/}
                {/* Menu component */}
                <ContrastButtons displayLabels={true}/> {/* <--- optional argument */}
            </div>
        )
    }
}

class ContrastButtons extends React.Component{
    render(){
        return(
            <div className="contrastButtons">
                {
                    this.props.displayLabels !== undefined 
                    && this.props.displayLabels === true ? 
                    <LabelledContrastButtons/> : <CompactContrastButtons/>}
            </div>
        )
    }
}

class LabelledContrastButtons extends React.Component{
    render(){
        return(
            <div className="labelledContrastButtons">
                <a href="#">Set to high contrast</a>
            </div>
        )
    }
}

class CompactContrastButtons extends React.Component{
    render(){
        return(
            <div className="compactContrastButtons">
                <a className="highContrastButton" href="#setToHightContrast">A</a>
                <a className="defaultContrastButton" href="#setToLowContrast">A</a>
            </div>
        )
    }
}
```

I hope this makes sense.

**Comment by brockfanning** [comment]
I'm following you there - but to help me understand the direction - how do you envision the user's repository looking?
1. Will the user's repository include _every_ component file?
2. Or only a subset of top-level component files?
3. Or only those component files the user wants to override?

**Comment by SavvasStephanides** [comment]
Good question. I'm thinking maybe some sort of a series of components that can be imported from npm in `node_modules` or something similar. Maybe it'll be some sort of "build your own UI" system where they'll have the freedom to either build a UI using our pre-built components, build their own from scratch and just use our new API to access the data, or keep the `open-sdg` frontend and just tweak it to access the data from our new API.

Of course this is just brainstorming at this point.

**Comment by brockfanning** [comment]
Definitely, this is good to think about. Thus far we have been allowing 2 levels of customization for users:

1. Config settings in `_config.yml` and in each indicator's metadata
2. The ability to override includes, layouts, and assets

It's worth mentioning that the latter item above is great for overall flexibility, but does complicate version upgrades (as you can see if you scan for the word "overrides" in our [release notes](https://open-sdg.readthedocs.io/en/latest/changelog/)).

Keeping both methods of customization would keep things familiar for our users. But that's not to say we shouldn't consider dropping the "override" approach, and try to meet our users needs with just the "config" approach.

Given that we're talking about a complete re-write, I'd propose a bit of "discovery" before diving into this. Perhaps a developer call so we can settle on what we like/dislike about the current platform. Then some surveying of our users to figure out their likes/dislikes as well. This should give us enough info to make informed decisions when we move onto the proofs-of-concept phase.

**Comment by SavvasStephanides** [comment]
> Given that we're talking about a complete re-write, I'd propose a bit of "discovery" before diving into this. Perhaps a developer call so we can settle on what we like/dislike about the current platform. Then some surveying of our users to figure out their likes/dislikes as well. 

I second this 100%. User research will give us a much better insight into what users need regarding the platform.

Regarding overriding, I personally found updating stuff to be quite confusing, having to juggle between `open-sdg` and `sdg-indicators` when updating the `scss` files. Not sure what other developers think of this though.

**Comment by lindamejia** [comment]
I am developer working on the City of LA platform and I completely agree with @SavvasStephanides, I found the overriding confusing as well. I think revamping the platform to React would make it much easier to work with. Is this still being considered?

**Comment by brockfanning** [comment]
@lindamejia Yes, this is definitely still up for consideration. We're on the verge of releasing a stable "version 1" of the platform, so I think it's safe to say that 1.0.0 will be Jekyll-based, as it is now. But this could make sense as a plan for version 2.0.0.

**Comment by SavvasStephanides** [comment]
@lindamejia 

Hello Linda! Awesome to hear other developers participating in the discussion and endorsing a component-based framework. As to whether this is being considered, this is part of a large pile of considerations on how to overall improve the SDG platform. There is no clear timeline on when this is going to be implemented as this is currently in the back burner along with other considerations. Still good to see endorsement from developers.

**Comment by LucyGwilliamAdmin** [comment]
Some benefits of this could be achieved by splitting up existing framework even further without adding new libraries. Next steps:

- Consider what could be split up
- Consider what most popular overrides are - could kick off discussion

**Comment by brockfanning** [comment]
I just started a separate discussion about Jekyll overrides here: #1478 

I would be greatly interested to hear from any developers familiar with both Open SDG and component-based frameworks like React or Vue. We need to start compiling a list of the benefits of switching to one of these frameworks. @lindamejia do you have any thoughts on benefits? And feedback from anyone else is welcome!
